nginx/src/charmlibs/nginx/_config.py

```diff
@@ -0,0 +1,638 @@
+    '^~',  # prefix match that disables further regex matching
+]
+
+
+@dataclass
+class NginxLocationConfig:
+    """Represents a `location` block in a Nginx configuration file.
+
+    For example::
+
+        NginxLocationConfig(
+            '/',
+            'foo',
+            backend_url="/api/v1"
+            headers={'a': 'b'},
+            modifier=EXACT,
+            is_grpc=True,
+            use_tls=True,
+        )
+
+    would result in the nginx config::
+
+        location = / {
+            set $backend grpcs://foo/api/v1;
+            grpc_pass $backend;
+            proxy_connect_timeout 5s;
+            proxy_set_header a b;
+        }
+    """
+
+    path: str
+    """The location path (e.g., '/', '/api') to match incoming requests."""
+    backend: str
+    """The name of the upstream service to route requests to (e.g. an `upstream` block)."""
+    backend_url: str = ''
+    """An optional URL path to append when forwarding to the upstream (e.g., '/v1')."""
+    headers: dict[str, str] = field(default_factory=lambda: cast('dict[str, str]', {}))
+    """Custom headers to include in the proxied request."""
+    modifier: _NginxLocationModifier = ''
+    """The Nginx location modifier."""
+    is_grpc: bool = False
+    """Whether to use gRPC proxying (i.e. `grpc_pass` instead of `proxy_pass`)."""
+    upstream_tls: bool | None = None
+    """Whether to connect to the upstream over TLS (e.g., https:// or grpcs://)
+    If None, it will inherit the TLS setting from the server block that the location is part of.
+    """
+
+
+@dataclass
+class NginxUpstream:
+    """Represents metadata needed to construct an Nginx `upstream` block."""
+
+    name: str
+    """Name of the upstream block."""
+    port: int
+    """Port number that all backend servers in this upstream listen on.
+
+    Our coordinators assume that all servers under an upstream share the same port.
+    """
+    group: str | None = None
+    """Group that this upstream belongs to.
+
+    Used for mapping multiple upstreams to a single group of backends (loadbalancing between all).
+    If you leave it None, this upstream will be routed to all available backends
+    (loadbalancing between them).
+    """
+
+
+def _is_ipv6_enabled() -> bool:
+    """Check if IPv6 is enabled on the container's network interfaces."""
+    try:
+        output = subprocess.run(
+            ['ip', '-6', 'address', 'show'], check=True, capture_output=True, text=True
+        )
+    except subprocess.CalledProcessError:
+        # if running the command failed for any reason, assume ipv6 is not enabled.
+        return False
+    return bool(output.stdout)
+
+
+class NginxConfig:
+    """NginxConfig.
+
+    To generate an Nginx configuration for a charm, instantiate the `NginxConfig` class with the
+      required inputs:
+
+    1. `server_name`: The name of the server (e.g. charm fqdn), which is used to identify the
+      server in Nginx configurations.
```

Comment:
```suggestion
    1. `server_name`: The name of the server (e.g. charm fqdn), which is used to identify the
       server in Nginx configurations.
```
I think we need one more space to appease Sphinx.